README.md
# node-triton-tags

This package defines and provides functions for parsing and validating "Triton
tags" -- special structure `triton.*` tags on instances (a.k.a. containers,
VMs) in [Joyent's Triton](https://www.joyent.com/triton). It exists mainly to
share Triton tag processing between the core
[sdc-docker](https://github.com/joyent/sdc-docker) and
[VMAPI](https://github.com/joyent/sdc-vmapi) services in Triton.


## Install

    npm install triton-tags


## Development Hooks

Before commiting be sure to, at least:

    make check      # lint and style checks
    make test       # run unit tests

A good way to do that is to install the stock pre-commit hook in your
clone via:

    make git-hooks

Also please run the full (longer) test suite (`make test`). See the next
section.


## Test suite

    make test


## License

MPL 2.0


CHANGES.md
# node-triton-tags changelog

## not yet released

(nothing yet)

## 1.3.0

- TRITON-1242: Add `triton.placement.exclude_virtual_servers` boolean tag. This
  is used to tell DAPI to filter out virtual servers (e.g. mockcloud).

## 1.2.2

- TRITON-755 Fix the release process (`make cutarelease`) to ensure pegjs files
  are built and included in the published files.

## 1.2.1

- TRITON-755: Fix missing peg.js generated files

## 1.2.0

- TRITON-755: Add support for 'triton.cmon.groups' tag. This tag is used for
  filtering the set of results returned by the CMON discover endpoint.

## 1.1.4

- DOCKER-1020: Add 'triton.network.public' tag

## 1.1.3

- joyent/node-triton-tags#2: allow DNS names to start with a number. Even
  though this is not recommended by the DNS RFCs, in practice it is used
  fairly commonly (including by a number of major websites)

## 1.1.2

- CNS-153: Upper case in services tags stopped working (another unplanned
  feature)

## 1.1.1

- CNS-152: allowing periods in service names: an unplanned feature

## 1.1.0

- CNS-147: new PEG parser for triton.cns.services tag, so this can be
  shared with triton-cns as well. Now supports ports for SRV records and
  extensible properties.

## 1.0.0

- DOCKER-736 Initial release.


test/unit.test.js
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

/*
 * Copyright 2018, Joyent, Inc.
 */

/*
 * Unit tests for "triton-tags"
 */

var format = require('util').format;
var test = require('tape');
var vasync = require('vasync');

var triton_tags = require('../');


// ---- test data

/*
 * `parseTritonTagStr` tests use key, str, val and err
 * `validateTritonTag` tests use key, val (or str if `val` isn't set) and
 *      errmsg (or `err` if errmsg isn't set)
 *
 */
var cases = [
    // Basics:
    {
        key: 'triton._test.string',
        str: 'astr',
        val: 'astr'
    },
    {
        key: 'triton._test.string',
        str: '',
        val: ''
    },
    {
        key: 'triton._test.boolean',
        str: 'true',
        val: true
    },
    {
        key: 'triton._test.boolean',
        str: 'false',
        val: false
    },
    {
        key: 'triton._test.number',
        str: '42',
        val: 42
    },

    // Some type failures:
    {
        key: 'triton._test.boolean',
        str: 'not a bool',
        /* JSSTYLED */
        err: /Triton tag "triton._test.boolean" value must be "true" or "false": "not a bool"/,
        /* JSSTYLED */
        errmsg: /Triton tag "triton._test.boolean" value must be a boolean: "not a bool"/
    },
    {
        key: 'triton._test.number',
        str: 'not a num',
        /* JSSTYLED */
        err: /Triton tag "triton._test.number" value must be a number: "not a num"/
    },
    {
        key: 'triton._test.number',
        str: '',
        /* JSSTYLED */
        err: /Triton tag "triton._test.number" value must be a number: ""/
    },

    // Unknown tag:
    {
        key: 'triton.unknown',
        str: '',
        /* JSSTYLED */
        err: /Unrecognized special triton tag \"triton.unknown\"/
    },

    // triton.cmon.groups
    {
        key: 'triton.cmon.groups',
        str: '',
        /* JSSTYLED */
        err: /invalid \"triton.cmon.groups\" tag: group name must be no less than 1 character and no greater than 100 characters/
    },
    {
        key: 'triton.cmon.groups',
        str: '&(--',
        /* JSSTYLED */
        err: /invalid \"triton.cmon.groups\" tag: groups must be strings comprised of letters, numbers, _, and -/
    },
    {
        key: 'triton.cmon.groups',
        /* JSSTYLED */
        str: '1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101',
        /* JSSTYLED */
        err: /invalid "triton.cmon.groups" tag: must contain less than or equal to 100 group strings/
    },
    {
        key: 'triton.cmon.groups',
        /* JSSTYLED */
        str: 'nifjdwhctkxbyhdqtwifkmehiedqocmmbguukckllseyirkedddrhmqmaemwfczfisvpmhunykccavpxqenpegbymhibsdazfmrrc',
        /* JSSTYLED */
        err: /invalid "triton.cmon.groups" tag: group name must be no less than 1 character and no greater than 100 characters/
    },
    {
        key: 'triton.cmon.groups',
        str: 'a,b,c,dupe,z,dupe,x',
        /* JSSTYLED */
        err: /invalid \"triton.cmon.groups\" tag: contains duplicate group dupe/
    },
    {
        key: 'triton.cmon.groups',
        str: 'foo-bar,bar_foo,disc',
        val: 'foo-bar,bar_foo,disc'
    },

    // triton.cns.disable
    {
        key: 'triton.cns.disable',
        str: 'true',
        val: true
    },
    {
        key: 'triton.cns.disable',
        str: 'false',
        val: false
    },
    {
        key: 'triton.cns.disable',
        str: 'booga',
        /* JSSTYLED */
        err: /Triton tag "triton.cns.disable" value must be "true" or "false": "booga"/,
        /* JSSTYLED */
        errmsg: /Triton tag "triton.cns.disable" value must be a boolean: "booga"/
    },

    {
        key: 'triton.cns.services',
        str: '',
        err: /Expected DNS name but end of input found/
    },
    {
        key: 'triton.cns.services',
        str: 'foobar',
        val: 'foobar'
    },
    {
        key: 'triton.cns.services',
        str: '_foobar',
        /* JSSTYLED */
        err: /Expected DNS name but "_" found/
    },
    {
        key: 'triton.cns.services',
        str: 'foobar:1234',
        val: 'foobar:1234'
    },
    {
        key: 'triton.cns.services',
        str: 'foobar,test',
        val: 'foobar,test'
    },
    {
        key: 'triton.cns.services',
        str: 'foobar.dev:123',
        val: 'foobar.dev:123'
    },
    {
        key: 'triton.cns.services',
        str: 'foobar:1234,test:1234',
        val: 'foobar:1234,test:1234'
    },
    {
        key: 'triton.cns.services',
        str: 'foobar:abcd',
        /* JSSTYLED */
        err: /Expected "=" but end of input found/
    },
    {
        key: 'triton.cns.services',
        str: 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa',
        err: /63 or fewer characters/
    },
    {
        key: 'triton.cns.services',
        str: 'foobar:123123123123',
        err: /must be within the range 1 - 65535/
    },
    {
        key: 'triton.cns.services',
        str: 'foobar:1234:priority=10',
        val: 'foobar:1234:priority=10'
    },
    {
        key: 'triton.cns.services',
        str: 'foobar:port=1234',
        val: 'foobar:port=1234'
    },
    {
        key: 'triton.cns.services',
        str: 'foobar:invalid=somevalue1',
        err: /not a valid property name/
    },
    {
        key: 'triton.cns.services',
        str: 'foobar:priority=aaaaaaa',
        err: /must be within the range 0 - 65535/
    },
    {
        key: 'triton.cns.services',
        str: '5foobar',
        val: '5foobar'
    }

    // TODO: triton.cns.reverse_ptr
];


// ---- tests

test('isTritonTag', function (t) {
    var isTritonTag = triton_tags.isTritonTag;

    [
        'triton.foo',
        'triton.cns.disable'
    ].forEach(function (key) {
        t.equal(isTritonTag(key), true, 'is a triton tag: ' + key);
    });

    [
        'Triton.foo',
        'cns.disable',
        ''
    ].forEach(function (key) {
        t.equal(isTritonTag(key), false, 'is not a triton tag: ' + key);
    });

    t.end();
});


test('parseTritonTagStr', function (t) {
    var parseTritonTagStr = triton_tags.parseTritonTagStr;

    vasync.forEachPipeline({
        inputs: cases,
        func: function testOneCase(c, next) {
            parseTritonTagStr(c.key, c.str, function (err, val) {
                var name = format('parseTritonTagStr(%j, %j)',
                    c.key, c.str);
                if (c.err) {
                    t.ok(err, name + ' (expect err)');
                    t.ok(c.err.exec(err.message), format(
                        'err.message matches %s: %j', c.err, err.message));
                    t.ok(val === undefined);
                } else {
                    t.ifErr(err, name);
                    t.equal(val, c.val, 'val');
                }
                next();
            });
        }
    }, function (err) {
        t.ifErr(err, 'parseTritonTagStr cases');
        t.end();
    });
});

test('validateTritonTag', function (t) {
    var validateTritonTag = triton_tags.validateTritonTag;

    vasync.forEachPipeline({
        inputs: cases,
        func: function testOneCase(c, next) {
            var val = (c.hasOwnProperty('val') ? c.val : c.str);
            var name = format('validateTritonTag(%j, %j)', c.key, val);
            var errmsg = validateTritonTag(c.key, val);
            if (c.errmsg || c.err) {
                t.ok(errmsg, name + ' (expect err)');
                t.ok((c.errmsg || c.err).exec(errmsg), format(
                    'errmsg matches %s: %j', (c.errmsg || c.err), errmsg));
            } else {
                t.ifErr(errmsg, name);
            }
            next();
        }
    }, function (err) {
        t.ifErr(err, 'validateTritonTag cases');
        t.end();
    });
});

var parser = require('../lib/cns-svc-tag');
test('cns services tag regressions', function (t) {
    /* CNS-152 */
    t.deepEqual(parser.parse('foo.dev,test.dev'), [
        { name: 'foo.dev' }, { name: 'test.dev' }
    ]);
    /* CNS-153 */
    t.deepEqual(parser.parse('someName'), [
        { name: 'somename' }
    ]);
    t.end();
});


lib/cmon-groups-tag.js
/*
 * Generated by PEG.js 0.10.0.
 *
 * http://pegjs.org/
 */

"use strict";

function peg$subclass(child, parent) {
  function ctor() { this.constructor = child; }
  ctor.prototype = parent.prototype;
  child.prototype = new ctor();
}

function peg$SyntaxError(message, expected, found, location) {
  this.message  = message;
  this.expected = expected;
  this.found    = found;
  this.location = location;
  this.name     = "SyntaxError";

  if (typeof Error.captureStackTrace === "function") {
    Error.captureStackTrace(this, peg$SyntaxError);
  }
}

peg$subclass(peg$SyntaxError, Error);

peg$SyntaxError.buildMessage = function(expected, found) {
  var DESCRIBE_EXPECTATION_FNS = {
        literal: function(expectation) {
          return "\"" + literalEscape(expectation.text) + "\"";
        },

        "class": function(expectation) {
          var escapedParts = "",
              i;

          for (i = 0; i < expectation.parts.length; i++) {
            escapedParts += expectation.parts[i] instanceof Array
              ? classEscape(expectation.parts[i][0]) + "-" + classEscape(expectation.parts[i][1])
              : classEscape(expectation.parts[i]);
          }

          return "[" + (expectation.inverted ? "^" : "") + escapedParts + "]";
        },

        any: function(expectation) {
          return "any character";
        },

        end: function(expectation) {
          return "end of input";
        },

        other: function(expectation) {
          return expectation.description;
        }
      };

  function hex(ch) {
    return ch.charCodeAt(0).toString(16).toUpperCase();
  }

  function literalEscape(s) {
    return s
      .replace(/\\/g, '\\\\')
      .replace(/"/g,  '\\"')
      .replace(/\0/g, '\\0')
      .replace(/\t/g, '\\t')
      .replace(/\n/g, '\\n')
      .replace(/\r/g, '\\r')
      .replace(/[\x00-\x0F]/g,          function(ch) { return '\\x0' + hex(ch); })
      .replace(/[\x10-\x1F\x7F-\x9F]/g, function(ch) { return '\\x'  + hex(ch); });
  }

  function classEscape(s) {
    return s
      .replace(/\\/g, '\\\\')
      .replace(/\]/g, '\\]')
      .replace(/\^/g, '\\^')
      .replace(/-/g,  '\\-')
      .replace(/\0/g, '\\0')
      .replace(/\t/g, '\\t')
      .replace(/\n/g, '\\n')
      .replace(/\r/g, '\\r')
      .replace(/[\x00-\x0F]/g,          function(ch) { return '\\x0' + hex(ch); })
      .replace(/[\x10-\x1F\x7F-\x9F]/g, function(ch) { return '\\x'  + hex(ch); });
  }

  function describeExpectation(expectation) {
    return DESCRIBE_EXPECTATION_FNS[expectation.type](expectation);
  }

  function describeExpected(expected) {
    var descriptions = new Array(expected.length),
        i, j;

    for (i = 0; i < expected.length; i++) {
      descriptions[i] = describeExpectation(expected[i]);
    }

    descriptions.sort();

    if (descriptions.length > 0) {
      for (i = 1, j = 1; i < descriptions.length; i++) {
        if (descriptions[i - 1] !== descriptions[i]) {
          descriptions[j] = descriptions[i];
          j++;
        }
      }
      descriptions.length = j;
    }

    switch (descriptions.length) {
      case 1:
        return descriptions[0];

      case 2:
        return descriptions[0] + " or " + descriptions[1];

      default:
        return descriptions.slice(0, -1).join(", ")
          + ", or "
          + descriptions[descriptions.length - 1];
    }
  }

  function describeFound(found) {
    return found ? "\"" + literalEscape(found) + "\"" : "end of input";
  }

  return "Expected " + describeExpected(expected) + " but " + describeFound(found) + " found.";
};

function peg$parse(input, options) {
  options = options !== void 0 ? options : {};

  var peg$FAILED = {},

      peg$startRuleFunctions = { tag: peg$parsetag },
      peg$startRuleFunction  = peg$parsetag,

      peg$c0 = ",",
      peg$c1 = peg$literalExpectation(",", false),
      peg$c2 = function(first, rest) {
      	var all = rest.map(function (n) { return (n[1]); });
      	all.unshift(first);
      	return (all);
      },
      peg$c3 = peg$otherExpectation("Group"),
      peg$c4 = /^[a-zA-Z0-9\-_]/,
      peg$c5 = peg$classExpectation([["a", "z"], ["A", "Z"], ["0", "9"], "-", "_"], false, false),
      peg$c6 = function() {
          return (text());
      },

      peg$currPos          = 0,
      peg$savedPos         = 0,
      peg$posDetailsCache  = [{ line: 1, column: 1 }],
      peg$maxFailPos       = 0,
      peg$maxFailExpected  = [],
      peg$silentFails      = 0,

      peg$result;

  if ("startRule" in options) {
    if (!(options.startRule in peg$startRuleFunctions)) {
      throw new Error("Can't start parsing from rule \"" + options.startRule + "\".");
    }

    peg$startRuleFunction = peg$startRuleFunctions[options.startRule];
  }

  function text() {
    return input.substring(peg$savedPos, peg$currPos);
  }

  function location() {
    return peg$computeLocation(peg$savedPos, peg$currPos);
  }

  function expected(description, location) {
    location = location !== void 0 ? location : peg$computeLocation(peg$savedPos, peg$currPos)

    throw peg$buildStructuredError(
      [peg$otherExpectation(description)],
      input.substring(peg$savedPos, peg$currPos),
      location
    );
  }

  function error(message, location) {
    location = location !== void 0 ? location : peg$computeLocation(peg$savedPos, peg$currPos)

    throw peg$buildSimpleError(message, location);
  }

  function peg$literalExpectation(text, ignoreCase) {
    return { type: "literal", text: text, ignoreCase: ignoreCase };
  }

  function peg$classExpectation(parts, inverted, ignoreCase) {
    return { type: "class", parts: parts, inverted: inverted, ignoreCase: ignoreCase };
  }

  function peg$anyExpectation() {
    return { type: "any" };
  }

  function peg$endExpectation() {
    return { type: "end" };
  }

  function peg$otherExpectation(description) {
    return { type: "other", description: description };
  }

  function peg$computePosDetails(pos) {
    var details = peg$posDetailsCache[pos], p;

    if (details) {
      return details;
    } else {
      p = pos - 1;
      while (!peg$posDetailsCache[p]) {
        p--;
      }

      details = peg$posDetailsCache[p];
      details = {
        line:   details.line,
        column: details.column
      };

      while (p < pos) {
        if (input.charCodeAt(p) === 10) {
          details.line++;
          details.column = 1;
        } else {
          details.column++;
        }

        p++;
      }

      peg$posDetailsCache[pos] = details;
      return details;
    }
  }

  function peg$computeLocation(startPos, endPos) {
    var startPosDetails = peg$computePosDetails(startPos),
        endPosDetails   = peg$computePosDetails(endPos);

    return {
      start: {
        offset: startPos,
        line:   startPosDetails.line,
        column: startPosDetails.column
      },
      end: {
        offset: endPos,
        line:   endPosDetails.line,
        column: endPosDetails.column
      }
    };
  }

  function peg$fail(expected) {
    if (peg$currPos < peg$maxFailPos) { return; }

    if (peg$currPos > peg$maxFailPos) {
      peg$maxFailPos = peg$currPos;
      peg$maxFailExpected = [];
    }

    peg$maxFailExpected.push(expected);
  }

  function peg$buildSimpleError(message, location) {
    return new peg$SyntaxError(message, null, null, location);
  }

  function peg$buildStructuredError(expected, found, location) {
    return new peg$SyntaxError(
      peg$SyntaxError.buildMessage(expected, found),
      expected,
      found,
      location
    );
  }

  function peg$parsetag() {
    var s0, s1, s2, s3, s4, s5;

    s0 = peg$currPos;
    s1 = peg$parsegroup();
    if (s1 !== peg$FAILED) {
      s2 = [];
      s3 = peg$currPos;
      if (input.charCodeAt(peg$currPos) === 44) {
        s4 = peg$c0;
        peg$currPos++;
      } else {
        s4 = peg$FAILED;
        if (peg$silentFails === 0) { peg$fail(peg$c1); }
      }
      if (s4 !== peg$FAILED) {
        s5 = peg$parsegroup();
        if (s5 !== peg$FAILED) {
          s4 = [s4, s5];
          s3 = s4;
        } else {
          peg$currPos = s3;
          s3 = peg$FAILED;
        }
      } else {
        peg$currPos = s3;
        s3 = peg$FAILED;
      }
      while (s3 !== peg$FAILED) {
        s2.push(s3);
        s3 = peg$currPos;
        if (input.charCodeAt(peg$currPos) === 44) {
          s4 = peg$c0;
          peg$currPos++;
        } else {
          s4 = peg$FAILED;
          if (peg$silentFails === 0) { peg$fail(peg$c1); }
        }
        if (s4 !== peg$FAILED) {
          s5 = peg$parsegroup();
          if (s5 !== peg$FAILED) {
            s4 = [s4, s5];
            s3 = s4;
          } else {
            peg$currPos = s3;
            s3 = peg$FAILED;
          }
        } else {
          peg$currPos = s3;
          s3 = peg$FAILED;
        }
      }
      if (s2 !== peg$FAILED) {
        peg$savedPos = s0;
        s1 = peg$c2(s1, s2);
        s0 = s1;
      } else {
        peg$currPos = s0;
        s0 = peg$FAILED;
      }
    } else {
      peg$currPos = s0;
      s0 = peg$FAILED;
    }

    return s0;
  }

  function peg$parsegroup() {
    var s0, s1, s2;

    peg$silentFails++;
    s0 = peg$currPos;
    s1 = [];
    if (peg$c4.test(input.charAt(peg$currPos))) {
      s2 = input.charAt(peg$currPos);
      peg$currPos++;
    } else {
      s2 = peg$FAILED;
      if (peg$silentFails === 0) { peg$fail(peg$c5); }
    }
    while (s2 !== peg$FAILED) {
      s1.push(s2);
      if (peg$c4.test(input.charAt(peg$currPos))) {
        s2 = input.charAt(peg$currPos);
        peg$currPos++;
      } else {
        s2 = peg$FAILED;
        if (peg$silentFails === 0) { peg$fail(peg$c5); }
      }
    }
    if (s1 !== peg$FAILED) {
      peg$savedPos = s0;
      s1 = peg$c6();
    }
    s0 = s1;
    peg$silentFails--;
    if (s0 === peg$FAILED) {
      s1 = peg$FAILED;
      if (peg$silentFails === 0) { peg$fail(peg$c3); }
    }

    return s0;
  }

  peg$result = peg$startRuleFunction();

  if (peg$result !== peg$FAILED && peg$currPos === input.length) {
    return peg$result;
  } else {
    if (peg$result !== peg$FAILED && peg$currPos < input.length) {
      peg$fail(peg$endExpectation());
    }

    throw peg$buildStructuredError(
      peg$maxFailExpected,
      peg$maxFailPos < input.length ? input.charAt(peg$maxFailPos) : null,
      peg$maxFailPos < input.length
        ? peg$computeLocation(peg$maxFailPos, peg$maxFailPos + 1)
        : peg$computeLocation(peg$maxFailPos, peg$maxFailPos)
    );
  }
}

module.exports = {
  SyntaxError: peg$SyntaxError,
  parse:       peg$parse
};


lib/index.js
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

/*
 * Copyright 2019, Joyent, Inc.
 */

/*
 * In Triton some per-VM configuration is controlled via special structured
 * tags on the VM's "tags" object. They are all prefixed with "triton.".
 * Let's call them "Triton tags".
 *
 * Triton tags are all optional.
 *
 * - `triton.cmon.groups` (string): Comma-separated list of CMON group strings
 *   used to filter the set of results returned by the CMON discover endpoint
 *   (e.g. Request that only VMs with tag 'triton.cmon.groups=api' be returned).
 * - `triton.cns.disable` (boolean): Can be set on a VM to tell the CNS service
 *   to not serve records for this VM.
 * - `triton.cns.services` (string): Comma-separated list of DNS-name strings
 *   for the CNS service.
 * - `triton.cns.reverse_ptr` (string): DNS reverse pointer for this VM. Used
 *   by the CNS service.
 * - `triton.network.public` (string): Set on a VM, used to specify the external
 *   network name a VM will use.
 * - `triton.placement.exclude_virtual_servers` (boolean): Set on a VM, used to
 *   tell DAPI to ignore virtual servers when provisioning (e.g. mockcloud).
 */

var assert = require('assert-plus');
var format = require('util').format;
var cmonGroupsTag = require('./cmon-groups-tag');
var cnsSvcTag = require('./cns-svc-tag');

/*
 * For now, using the more limited labels allowed by RFC1123. RFC2181 supercedes
 * 1123, but the broader range of characters can sometimes cause problems with
 * other systems (e.g. see the underscore in RFC5321).
 */
var DNS_NAME_RE = /^[a-z0-9][a-z0-9\-]{0,62}(?:\.[a-z0-9][a-z0-9\-]{0,62})*$/i;


/*
 * The value 100 is somewhat arbitrary, but was chosen as a limit on the number
 * of gropus and length of group names. The reasoning is that it is likely high
 * enough to not be hit, but low enough to not allow massive group counts and
 * group names that could potentially chew up DRAM in the CMON cache.
 */
var CMON_MAX_GROUPS = 100;
var CMON_MAX_GROUP_LEN = 100;


var typeFromKey = {
    'triton.cmon.groups': 'string',
    'triton.cns.services': 'string',
    'triton.cns.disable': 'boolean',
    'triton.cns.reverse_ptr': 'string',
    'triton.network.public': 'string',
    'triton.placement.exclude_virtual_servers': 'boolean',

    // Internal testing Triton tags
    'triton._test.string': 'string',
    'triton._test.number': 'number',
    'triton._test.boolean': 'boolean'
};

/*
 * Validator functions take a `val` (already checked to be of the appropriate
 * type per `typeFromKey`) and return an error message if invalid. If valid,
 * null is returned.
 */
var validatorFromKey = {
    'triton.cmon.groups': function validateTritonCmonGroups(val) {
        assert.string(val, 'val');

        var groups;
        try {
            groups = cmonGroupsTag.parse(val);
        } catch (e) {
            return format(
                'invalid "triton.cmon.groups" tag: groups must be strings ' +
                'comprised of letters, numbers, _, and -');
        }

        assert.arrayOfString(groups);

        if (groups.length < 1) {
            return format(
                'invalid "triton.cmon.groups" tag: must contain at least ' +
                'one valid group string');
        }

        if (groups.length > CMON_MAX_GROUPS) {
            return format(
                'invalid "triton.cmon.groups" tag: must contain less than ' +
                'or equal to 100 group strings');
        }

        var tagCount = {};
        for (var i = 0; i < groups.length; ++i) {
            var group = groups[i];
            assert.string(group);

            if (tagCount[group] !== 1) {
                tagCount[group] = 1;
            } else {
                return format('invalid "triton.cmon.groups" tag: contains ' +
                    'duplicate group %s', group);
            }

            if (group.length < 1 || group.length > CMON_MAX_GROUP_LEN) {
                return format(
                    'invalid "triton.cmon.groups" tag: group name must be no ' +
                    'less than 1 character and no greater than 100 characters');
            }
        }

        return null;
    },

    'triton.cns.services': function validateTritonCnsServices(val) {
        assert.string(val, 'val');

        var svcs;
        try {
            svcs = cnsSvcTag.parse(val);
        } catch (e) {
            return format(
                'invalid "triton.cns.services" tag: %s', e.message);
        }
        assert.arrayOfObject(svcs);
        if (svcs.length < 1) {
            return format(
                'invalid "triton.cns.services" tag: must contain at least ' +
                'one valid service');
        }

        for (var i = 0; i < svcs.length; ++i) {
            var svc = svcs[i];
            assert.string(svc.name);
            if (svc.name.length < 1 || svc.name.length > 63) {
                return format(
                    'invalid "triton.cns.services" tag: service DNS name ' +
                    '"%s" must be 63 or fewer characters',
                    svc.name);
            }
            if (svc.port !== undefined) {
                svc.port = parseInt(svc.port, 10);
                if (isNaN(svc.port) || svc.port < 1 || svc.port > 65535) {
                    return format(
                        'invalid "triton.cns.services" tag: service port ' +
                        'number for %s must be within the range 1 - 65535',
                        svc.name);
                }
            }
            var keys = Object.keys(svc);
            for (var j = 0; j < keys.length; ++j) {
                if (keys[j] === 'name' || keys[j] === 'port') {
                    continue;

                } else if (keys[j] === 'priority' || keys[j] === 'weight') {
                    svc[keys[j]] = parseInt(svc[keys[j]], 10);

                    if (isNaN(svc[keys[j]]) || svc[keys[j]] < 0 ||
                        svc[keys[j]] > 65535) {

                        return format(
                            'invalid "triton.cns.services" tag: service ' +
                            '%s for %s must be within the range 0 - 65535',
                            keys[j], svc.name);
                    }

                    continue;

                } else {
                    return format(
                        'invalid "triton.cns.services" tag: service property ' +
                        '"%s" is not a valid property name', keys[j]);
                }
            }
        }

        return null;
    },

    'triton.cns.disable': function validateTritonCnsDisable(val) {
        assert.bool(val, 'val');

        // In other words, basically a no-op.
        return null;
    },

    'triton.cns.reverse_ptr': function validateTritonCnsReversePtr(val) {
        assert.string(val, 'val');

        if (val.length > 255 || !val.match(DNS_NAME_RE)) {
            return format(
                'invalid "triton.cns.reverse_ptr" tag: "%s" is not DNS safe',
                val);
        }

        return null;
    },

    'triton.network.public': function validateTritonNetworkPublic(val) {
        assert.string(val, 'val');
        return null;
    },

    'triton.placement.exclude_virtual_servers':
            function validateTritonExcludeVirtualServersBoolean(val) {
        assert.bool(val, 'val');
        return null;
    },

    /*
     * Triton tags used for internal testing.
     */
    'triton._test.string': function validateTritonTestString(val) {
        assert.string(val, 'val');
        return null;
    },
    'triton._test.number': function validateTritonTestNumber(val) {
        assert.number(val, 'val');
        return null;
    },
    'triton._test.boolean': function validateTritonTestBoolean(val) {
        assert.bool(val, 'val');
        return null;
    }
};



// --- exports

/*
 * Return true if the given key uses the Triton tag prefix.
 * Note that it still might not be one of the specific defined tags.
 */
function isTritonTag(key) {
    assert.string(key, 'key');
    const TRITON_TAG_PREFIX = 'triton.';
    return (key.substr(0, TRITON_TAG_PREFIX.length) === TRITON_TAG_PREFIX);
}


/**
 * Parse a Triton tag from the given string value.
 *
 * @param {String} key: The Triton tag key/name.
 * @param {String} str: The tag value to parse.
 * @param {Function} cb: `function (err, val)`
 *      On success, `err` is null and `val` is the parsed and validated
 *      tag value. On failure `err` is an Error instance.
 */
function parseTritonTagStr(key, str, cb) {
    assert.string(key, 'key');
    assert.string(str, 'str');
    assert.func(cb, 'cb');

    // Ensure it is a known triton tag.
    var type = typeFromKey[key];
    if (type === undefined) {
        cb(new Error(format('Unrecognized special triton tag "%s"', key)));
        return;
    }

    // Convert from string to value of appropriate type.
    var val;
    switch (type) {
    case 'string':
        val = str;
        break;
    case 'boolean':
        if (str === 'true') {
            val = true;
        } else if (str === 'false') {
            val = false;
        } else {
            cb(new Error(format(
                'Triton tag "%s" value must be "true" or "false": %j',
                key, str)));
            return;
        }
        break;
    case 'number':
        val = Number(str);
        if (str.length === 0 || /* Guard against `Number('') === 0` wat. */
            isNaN(val))
        {
            cb(new Error(format('Triton tag "%s" value must be a number: %j',
                key, str)));
            return;
        }
        break;
    default:
        throw new Error('unexpected Triton tag type: ' + type);
    }

    // Validate.
    var validator = validatorFromKey[key];
    assert.func(validator, 'validator for tag ' + key);
    var errmsg = validator(val);

    if (errmsg) {
        cb(new Error(errmsg));
    } else {
        cb(null, val);
    }
}


/**
 * Validate the given Triton tag key and value.
 *
 * This differs from `parseTritonTagStr` in that the `val` argument should
 * already be of the correct type, and no value is returned.
 *
 * @param {String} key
 * @param {String|Number|Boolean} val
 * @returns {null|String} If valid, `null` is returned. Otherwise an error
 *      message is returned.
 */
function validateTritonTag(key, val) {
    assert.string(key, 'key');

    // Check type.
    var expectedType = typeFromKey[key];
    var actualType = typeof (val);
    if (expectedType === undefined) {
        return format('Unrecognized special triton tag "%s"', key);
    } else if (expectedType !== actualType) {
        return format('Triton tag "%s" value must be a %s: %j (%s)',
            key, expectedType, val, actualType);
    }

    // Validate.
    var validator = validatorFromKey[key];
    assert.func(validator, 'validator for tag ' + key);
    return validator(val);
}



module.exports = {
    isTritonTag: isTritonTag,
    parseTritonTagStr: parseTritonTagStr,
    validateTritonTag: validateTritonTag
};
